source/common/upstream/original_dst_cluster.cc

```diff
@@ -166,48 +167,76 @@ OriginalDstCluster::OriginalDstCluster(
-    const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
-    HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
-    all_hosts->emplace_back(host);
-    priority_set_.updateHosts(0,
-                              HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
-                              {}, {std::move(host)}, {}, absl::nullopt);
+  std::string address = host->address()->asString();
+  HostMultiMapSharedPtr new_host_map = std::make_shared<HostMultiMap>(*getCurrentHostMap());
+  auto it = new_host_map->find(address);
+  if (it != new_host_map->end()) {
+    it->second->hosts_.push_back(host);
+  } else {
+    new_host_map->emplace(address, std::make_shared<HostsForAddress>(host));
   }
+  ENVOY_LOG(debug, "addHost() adding {} {}.", *host, address);
+  setHostMap(new_host_map);
+
+  // Given the current config, only EDS clusters support multiple priorities.
+  ASSERT(priority_set_.hostSetsPerPriority().size() == 1);
+  const auto& first_host_set = priority_set_.getOrCreateHostSet(0);
+  HostVectorSharedPtr all_hosts(new HostVector(first_host_set.hosts()));
+  all_hosts->emplace_back(host);
+  priority_set_.updateHosts(0,
+                            HostSetImpl::partitionHosts(all_hosts, HostsPerLocalityImpl::empty()),
+                            {}, {std::move(host)}, {}, absl::nullopt);
 }
 
 void OriginalDstCluster::cleanup() {
   HostVectorSharedPtr keeping_hosts(new HostVector);
   HostVector to_be_removed;
-  ENVOY_LOG(trace, "Stale original dst hosts cleanup triggered.");
+  absl::flat_hash_set<absl::string_view> removed_addresses;
   auto host_map = getCurrentHostMap();
   if (!host_map->empty()) {
     ENVOY_LOG(trace, "Cleaning up stale original dst hosts.");
-    for (const auto& pair : *host_map) {
-      const std::string& addr = pair.first;
-      const HostSharedPtr& host = pair.second;
-      if (host->used()) {
-        ENVOY_LOG(trace, "Keeping active host {}.", addr);
-        keeping_hosts->emplace_back(host);
-        host->used(false); // Mark to be removed during the next round.
+    for (const auto& [addr, hosts] : *host_map) {
+      // Address is kept in the cluster if either of the two things happen:
+      // 1) a host has been recently selected for the address; 2) none of the
+      // hosts are currently in any of the connection pools.
+      // The set of hosts for a single address are treated as a unit.
+      bool keep = false;
+      if (hosts->used_) {
+        keep = true;
+        hosts->used_ = false; // Mark to be removed during the next round.
+      } else if (Runtime::runtimeFeatureEnabled(
+                     "envoy.reloadable_features.original_dst_rely_on_idle_timeout")) {
+        if (hosts->host_->used()) {
+          keep = true;
+        } else {
+          for (const auto& host : hosts->hosts_) {
+            if (host->used()) {
+              keep = true;
+              break;
+            }
+          }
+        }
```

Comment:
Can you add comments here on why we need to check the top level and all the other ones?

Reply:
Done.